### ISSUE TITLE
Is a bug or a feature? Something about `BooleanAccuracy`

### ISSUE THREAD

**Issue by WrRan** [allenai/allennlp#2228]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.7
 - AllenNLP version: 0.8.1-unreleased
 - PyTorch version: 1.0.0

**Question**
There is a test case producing errors:
```python
import torch

from allennlp.common.testing import AllenNlpTestCase
from allennlp.training.metrics import BooleanAccuracy


class BooleanAccuracyTest(AllenNlpTestCase):
    def test_accuracy_computation(self):
        accuracy = BooleanAccuracy()
        predictions = torch.Tensor([[0, 1],
                                    [2, 3],
                                    [4, 5],
                                    [6, 7]])
        targets = torch.Tensor([[0, 1],
                                [2, 2],
                                [4, 5],
                                [7, 7]])

        accuracy.reset()
        mask = torch.ones(4, 2)
        mask[1] *= 0.
        mask[3] *= 0
        accuracy(predictions, targets, mask)
        assert accuracy._correct_count == 2.
        assert accuracy._total_count == 2.
        assert accuracy.get_metric() == 2. / 2

        accuracy.reset()
        mask = torch.zeros(4, 2)
        accuracy(predictions, targets, mask)
        assert accuracy._correct_count == 0.
        assert accuracy._total_count == 0.
        self.assertRaises(ZeroDivisionError, accuracy.get_metric)
```

Run this test case, and you will get an error as follow:
```txt
...
    assert accuracy._correct_count == 2.
AssertionError
...
```

But I do not know whether it is right to use `mask` as above. So is a bug or a misuse?

**Comment by joelgrus** [comment]
I think your test is wrong, `correct_count` should be 4 here. `predictions[0]` and `predictions[2]` are both exactly correct, and `predictions[1]` and `predictions[3]` are both totally masked (which means they agree with the target in all non-masked positions), so they're both correct too.

**Comment by matt-gardner** [comment]
Isn't it 4 because there are two values in each batch? Masked values should definitely not be counted in the metrics.

**Comment by joelgrus** [comment]
the code is somewhat confusing, but it looks like it's actually checking for each of the `batch_size` predictions "is it the case that every non-masked prediction is equal to the corresponding target"

so it's counting the number of totally correct *prediction tensors*. here there are 4 prediction tensors, two are correct, and two are fully masked.

I don't know if that's what `BooleanAccuracy` is *supposed to* do, but that seems to be what the code is doing

**Comment by joelgrus** [comment]
but if you look at the code it's very explicit that `count` is the `batch_size` dimension

**Comment by matt-gardner** [comment]
Ok, thanks for the description, I'm on a phone and didn't look at the code. That behavior sounds like a bug to me.

**Comment by joelgrus** [comment]
it should be an easy fix, just so I'm clear, the desired behavior is 

(# of non-masked tensor entries that are correct) / (# of non-masked tensor entries)

?

(that's what I expected when I looked at the code, but it was so explicitly doing something else that I figured it was intentional)

**Comment by matt-gardner** [comment]
Yeah, that's what I would expect the behavior to be.  And either way it should be well-documented what the intended behavior is - completely masked instances are an edge case that needs documentation.

**Comment by joelgrus** [comment]
ok, after discussing with markn, I think the current "all or nothing" behavior is correct, but the way it handles completely masked instances is not, so I'll make a PR for that and make the docs clearer
